File: Medium/Python/59. Spiral Matrix II.py
class Solution(object):
    def generateMatrix(self, n):
        mat = [[0] * n for _ in range(n)]

        temp = 1
        row = 0
        rowEnd = n - 1
        col = 0
        colEnd = n - 1

        while temp <= n * n:
            # Shift from left to right
            for i in range(col, colEnd + 1):
                mat[row][i] = temp
                temp += 1
            row += 1

            # Shift from top to bot
            for i in range(row, rowEnd + 1):
                mat[i][colEnd] = temp
                temp += 1
            colEnd -= 1

            # Shift from right to left
            for i in range(colEnd, col - 1, -1):
                mat[rowEnd][i] = temp
                temp += 1
            rowEnd -= 1

            # Shift from bot to top
            for i in range(rowEnd, row - 1, -1):
                mat[i][col] = temp
                temp += 1
            col += 1

        return mat

File: Medium/Java/59. Spiral Matrix II.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] mat = new int[n][n];

        int temp = 1;
        int row = 0;
        int rowEnd = n - 1;
        int col = 0;
        int colEnd = n - 1;

        while(temp <= n * n){
            // Shift from left to right
            for(int i = col; i <= colEnd; i++){
                mat[row][i] = temp;
                temp++;
            }
            row++;

            // Shift from top to bot
            for(int i = row; i <= rowEnd; i++){
                mat[i][colEnd] = temp;
                temp++;
            }
            colEnd--;

            // Shift from right to left
            for(int i = colEnd; col <= i; i--){
                mat[rowEnd][i] = temp;
                temp++;
            }
            rowEnd--;

            // Shift from bot to top
            for(int i = rowEnd; row <= i; i--){
                mat[i][col] = temp;
                temp++;
            }
            col++;
        }

        return mat;
    }
}
